### ISSUE TITLE
type Date not checking for undefined causing generic failure

### ISSUE THREAD

**Issue by winfinit** [Automattic/mongoose#2833]
if field is defined in schema with type of Date, and value is passed for that fields as undefined, failure doesn't propagate to CastError, instead failing with TypeError when toString method is attempted against undefined value, causing obscure error as such:

```
TypeError: Cannot read property 'toString' of undefined)
```

instead of expected:

```
CastError: Cast to date failed for value "undefined" at path "myFieldName"
```

**Comment by winfinit** [comment]
pull request submitted:
https://github.com/Automattic/mongoose/pull/2834

**Comment by winfinit** [comment]
snippet reproducing this issue:

```javascript
var mongoose = require('mongoose');
var mySchema = mongoose.Schema({date: { type: Date }});
var myModel = mongoose.model('mySchema', mySchema);

var myObject = new myModel();
myObject.update({date: undefined});
```

output:
```
TypeError: Cannot read property 'toString' of undefined
...
```

new output with provided patch
```
CastError: Cast to date failed for value "undefined" at path "date"
...
```

**Comment by vkarpov15** [comment]
This only affects 3.8.x, 4.0 fixes this thanks to #2272, so #2834 is unnecessary. I'll backport it to 3.8 though
